src/Symfony/Component/Security/Core/Role/RoleHierarchy.php

```diff
@@ -65,9 +65,10 @@ protected function buildRoleMap()
                 }
 
                 $visited[] = $role;
-                $this->map[$main] = array_unique(array_merge($this->map[$main], $this->hierarchy[$role]));
+                $this->map[$main] = array_merge($this->map[$main], $this->hierarchy[$role]);
                 $additionalRoles = array_merge($additionalRoles, array_diff($this->hierarchy[$role], $visited));
             }
+            $this->map[$main] = array_unique($this->map[$main]);
         }
```

Comment:
I'd be interested to see how this would impact performance additionally if both `array_merge`s are moved outwards